sys/drv/hidkbd.c

```diff
@@ -0,0 +1,208 @@
+/*
+ * USB HID keyboard driver.
+ *
+ * Relevant documents:
+ *
+ * - Device Class Definition for Human Interface Devices (HID) Version 1.11,
+ *   5/27/01:
+ *     https://www.usb.org/sites/default/files/hid1_11.pdf
+ */
+#define KL_LOG KL_DEV
+#include <sys/devclass.h>
+#include <sys/device.h>
+#include <sys/errno.h>
+#include <dev/evdev.h>
+#include <sys/klog.h>
+#include <sys/libkern.h>
+#include <sys/sched.h>
+#include <dev/usb.h>
+
+#define HIDKBD_NMODKEYS 8
+#define HIDKBD_NKEYCODES 6
+
+/* HID keyboard input report. */
+typedef struct hidkbd_in_report {
+  uint8_t modkeys;
+  uint8_t reserved;
+  uint8_t keycodes[HIDKBD_NKEYCODES];
+} __packed hidkbd_in_report_t;
+
+/* HID keyboard software context. */
+typedef struct hidkbd_state {
+  hidkbd_in_report_t prev_report; /* the most recent report */
+  thread_t *thread;               /* scancode gathering thread */
+  evdev_dev_t *evdev;             /* corresponding evdev device */
+} hidkbd_state_t;
+
+/*
+ * Evdev handling functions.
+ */
+
+static void hidkbd_init_evdev(device_t *dev) {
+  hidkbd_state_t *hidkbd = dev->state;
+  evdev_dev_t *evdev = evdev_alloc();
+
+  /* Set the basic evdev parameters. */
+  usb_device_t *udev = usb_device_of(dev);
+  evdev_set_name(evdev, "HID keyboard");
+  evdev_set_id(evdev, BUS_USB, udev->vendor_id, udev->product_id, 0);
+
+  /* `EV_SYN` is required for every evdev device. */
+  evdev_support_event(evdev, EV_SYN);
+  evdev_support_event(evdev, EV_KEY);
+
+  evdev_support_all_hidkbd_keys(evdev);
+
+  /* NOTE: we assume that the HID keyboard supports hardware key repetiton,
+   * which isn't said to always be true. */
+  evdev_support_event(evdev, EV_REP);
+
+  /* Bind `evdev` to `hidkbd` for future references. */
+  hidkbd->evdev = evdev;
+
+  /* Finally, register the device in the file system. */
+  evdev_register(evdev);
+}
+
+/*
+ * Report processing functions.
+ */
+
+static void hidkbd_push_evdev_event(hidkbd_state_t *hidkbd,
+                                    uint8_t hidkbd_keycode, uint32_t value) {
+  uint16_t evdev_keycode = evdev_hid2key(hidkbd_keycode);
+  if (evdev_keycode == KEY_RESERVED)
+    return;
+  evdev_push_key(hidkbd->evdev, evdev_keycode, value);
+  evdev_sync(hidkbd->evdev);
+}
+
+static void hidkbd_process_modkeys(hidkbd_state_t *hidkbd, uint8_t modkeys) {
```

Comment:
Shouldn't you use output report to light up special keys like _caps lock_ and so on?